lib/Basics/expected.h

```diff
@@ -0,0 +1,215 @@
+#pragma once
+#include <stdexcept>
+#include <type_traits>
+#include "Assertions/Assert.h"
+
+namespace arangodb {
+template<typename T>
+struct expected {
+  expected() noexcept {}
+
+  template<typename... Args>
+  explicit expected(std::in_place_t, Args&&... args) noexcept(
+      std::is_nothrow_constructible_v<T, Args...>) requires
+      std::constructible_from<T, Args...> {
+    new (&_value) T(std::forward<Args>(args)...);
+    _state = kValue;
+  }
+
+  explicit expected(std::exception_ptr ex) noexcept
+      : _exception(std::move(ex)), _state(kException) {
+    static_assert(std::is_nothrow_move_constructible_v<std::exception_ptr>);
+  }
+
+  expected(expected const& other) noexcept(
+      std::is_nothrow_copy_constructible_v<T>) requires
+      std::is_copy_constructible_v<T> {
+    if (other._state == kValue) {
+      new (&_value) T(other._value);
+      _state = kValue;
+    } else if (other._state == kException) {
+      static_assert(std::is_nothrow_copy_constructible_v<std::exception_ptr>);
+      new (&_exception) std::exception_ptr(other._exception);
+      _state = kException;
+    }
+  }
+
+  expected(expected&& other) noexcept(
+      std::is_nothrow_move_constructible_v<T>) requires
+      std::move_constructible<T> {
+    if (other._state == kValue) {
+      new (&_value) T(std::move(other._value));
+      _state = kValue;
+    } else if (other._state == kException) {
+      static_assert(std::is_nothrow_move_constructible_v<std::exception_ptr>);
+      new (&_exception) std::exception_ptr(std::move(other._exception));
+      _state = kException;
+    }
+  }
```

Comment:
Shouldn't we reset other here?

Reply:
debatable. What state should a moved-from `expected<T>` have? Always `empty`? Do we want to keep the moved-from `T` and `std::exception_ptr` ?

Reply:
optional move constructor has a similar behavior: "does not make other empty: a moved-from std::optional still contains a value, but the value itself is moved from"